nix-parser/src/parser/expr/func.rs

```diff
@@ -31,6 +31,37 @@ fn simple(input: Tokens) -> IResult<Partial<FnDeclSimple>> {
 }
 
 fn formals(input: Tokens) -> IResult<Partial<FnDeclFormals>> {
+    // Definition:
+    // a leading token sequence B of a non-terminal A is a production rule of B
+    // such that if a non-terminal C satisfies the condition that all words in its
+    // language L(C) has a prefix as a word in L(B), then L(A) = L(C), ie. A and C
+    // produces the same set of words.
+    // Note: unfortunately computing a leading token sequence of any non-terminal in any grammar
+    // is undecidable, but luckily we know one for this.
+
+    // Quickly look for a leading token sequence, because recovery with partial
+    // parsing result may produces invalid parsing tree on the rest of the tokens
+    // if there is no sequence of valid leading tokens
+
+    // According to the grammar, the leading token sequence will be
+    // left-curly-bracket ( right-curly-bracket colon / ellipsis / ident ( question / comma / right-curly-bracket ))
```

Comment:
This could also speed up the parsing of large attrset files like `yarn.nix`, too, as @trha mentioned to me before.

Reply:
Ah, I see. Yeah, I can confirm the numbers. This benchmark is quite cool, by the way. :+1:

Reply:
Thanks a lot! I want to expand the benchmark and make it a bit better and more accurate with time. I recently added a throughput measurement (in MiB of text parsed per second) to the benchmark, which has been proving useful.